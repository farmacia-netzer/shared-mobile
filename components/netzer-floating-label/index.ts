export * from './netzer-floating-label.component';
