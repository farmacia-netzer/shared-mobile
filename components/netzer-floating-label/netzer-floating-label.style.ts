import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  floatingLabel: {
    position: 'absolute'
  }
});
