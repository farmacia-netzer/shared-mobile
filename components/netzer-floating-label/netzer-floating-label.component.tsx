import React, { useEffect, Fragment } from 'react';
import Animated, { useSharedValue, withTiming, useAnimatedStyle } from 'react-native-reanimated';

import { isEmpty } from '##services/util/utils';
import { styles } from './netzer-floating-label.style';
import { FONT_SIZE } from '##theme/typography.constant';

interface NetzerFloatingLabelProp {
  isVisible?: boolean;
  isFocused?: boolean;
  value?: string | null;
  testID?: string;
  placeholder: string;
  placeholderTextColor?: string;
  inputStyle?:
    | {
        paddingTop?: number;
        paddingVertical?: number;
        fontSize?: number;
      }
    | {
        paddingTop?: number;
        paddingVertical?: number;
        fontSize?: number;
      };
}

export const NetzerFloatingLabel: React.FC<NetzerFloatingLabelProp> = ({
  isVisible = true,
  isFocused = false,
  placeholder = '',
  placeholderTextColor = '',
  value = '',
  inputStyle,
  testID
}: NetzerFloatingLabelProp) => {
  const DEFAULT_DURATION = 150;

  const DEFAULT_FONT_SIZE = inputStyle?.fontSize || FONT_SIZE.NORMAL;
  const DEFAULT_FONT_SHIRINK_SIZE = Math.floor(DEFAULT_FONT_SIZE * 0.86);
  const DEFAULT_LABEL_POSITION = inputStyle?.paddingTop || inputStyle?.paddingVertical || 0;

  const floatingLabelPosition = useSharedValue(DEFAULT_LABEL_POSITION);
  const floatingLabelFontSize = useSharedValue(DEFAULT_FONT_SIZE);

  useEffect(() => {
    if (isFocused) {
      floatingLabelPosition.value = 0;
      floatingLabelFontSize.value = DEFAULT_FONT_SHIRINK_SIZE;
    } else {
      if (isEmpty(value)) {
        floatingLabelPosition.value = DEFAULT_LABEL_POSITION;
        floatingLabelFontSize.value = DEFAULT_FONT_SIZE;
      } else {
        floatingLabelPosition.value = 0;
        floatingLabelFontSize.value = DEFAULT_FONT_SHIRINK_SIZE;
      }
    }
  }, [
    isFocused,
    value,
    floatingLabelPosition,
    floatingLabelFontSize,
    DEFAULT_LABEL_POSITION,
    DEFAULT_FONT_SIZE,
    DEFAULT_FONT_SHIRINK_SIZE
  ]);

  const floatingLabelStyle = useAnimatedStyle(() => {
    return {
      top: withTiming(floatingLabelPosition.value, { duration: DEFAULT_DURATION })
    };
  });

  const floatingLabelTextStyle = useAnimatedStyle(() => {
    return {
      fontSize: withTiming(floatingLabelFontSize.value, { duration: DEFAULT_DURATION }),
      color: placeholderTextColor
    };
  });

  return (
    <Fragment>
      {isVisible ? (
        <Animated.View
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          animation-config={{ useNativeDriver: true }}
          style={[styles.floatingLabel, floatingLabelStyle]}
          testID={testID + '-CONTAINER'}
        >
          <Animated.Text
            animation-config={{ useNativeDriver: true }}
            testID={testID + '-TEXT'}
            style={floatingLabelTextStyle}
          >
            {placeholder}
          </Animated.Text>
        </Animated.View>
      ) : null}
    </Fragment>
  );
};
