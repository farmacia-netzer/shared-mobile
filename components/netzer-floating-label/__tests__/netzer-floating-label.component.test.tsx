import React from 'react';
import { render } from '@testing-library/react-native';

import { NetzerFloatingLabel } from '../netzer-floating-label.component';
import { setMockRequestAnimationFrame } from '##__test__/reanimated-helper';

describe('NetzerFloatingLabel', () => {
  const mock_font_size = 14;
  const mock_top_position = 15;
  const mockPlaceholder = 'test placeholder';
  const mockPlaceholderTextColor = '#767676';
  const mockInputStyle = {
    paddingTop: mock_top_position,
    fontSize: mock_font_size
  };
  const mockTestID = 'TEST_ID';

  beforeEach(() => {
    setMockRequestAnimationFrame();
  });

  it('Renders properly when inVisible', () => {
    const { toJSON } = render(<NetzerFloatingLabel placeholder={mockPlaceholder} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders properly if value sets null and isFocused sets false', () => {
    const { toJSON } = render(
      <NetzerFloatingLabel
        isVisible={true}
        isFocused={false}
        placeholder={mockPlaceholder}
        placeholderTextColor={mockPlaceholderTextColor}
        inputStyle={mockInputStyle}
        testID={mockTestID}
        value={null}
      />
    );
    // jest.runAllTimers();
    expect(toJSON()).toMatchSnapshot();
  });
});
