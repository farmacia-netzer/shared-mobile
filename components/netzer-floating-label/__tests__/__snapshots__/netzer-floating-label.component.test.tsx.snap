// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetzerFloatingLabel Renders properly when inVisible 1`] = `
<View
  accessibilityElementsHidden={true}
  animation-config={
    Object {
      "useNativeDriver": true,
    }
  }
  importantForAccessibility="no-hide-descendants"
  style={
    Array [
      Object {
        "position": "absolute",
      },
      Object {
        "top": 0,
      },
    ]
  }
  testID="undefined-CONTAINER"
>
  <Text
    animation-config={
      Object {
        "useNativeDriver": true,
      }
    }
    style={
      Object {
        "color": "",
        "fontSize": 14,
      }
    }
    testID="undefined-TEXT"
  >
    test placeholder
  </Text>
</View>
`;

exports[`NetzerFloatingLabel renders properly if value sets null and isFocused sets false 1`] = `
<View
  accessibilityElementsHidden={true}
  animation-config={
    Object {
      "useNativeDriver": true,
    }
  }
  importantForAccessibility="no-hide-descendants"
  style={
    Array [
      Object {
        "position": "absolute",
      },
      Object {
        "top": 15,
      },
    ]
  }
  testID="TEST_ID-CONTAINER"
>
  <Text
    animation-config={
      Object {
        "useNativeDriver": true,
      }
    }
    style={
      Object {
        "color": "#767676",
        "fontSize": 14,
      }
    }
    testID="TEST_ID-TEXT"
  >
    test placeholder
  </Text>
</View>
`;
